libwallet/ext/rate_source.go

```diff
@@ -0,0 +1,923 @@
+// Copyright (c) 2019-2021, The Decred developers
+// Copyright (c) 2023, The Cryptopower developers
+// See LICENSE for details.
+
+package ext
+
+import (
+	"bytes"
+	"compress/flate"
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/crypto-power/cryptopower/libwallet/utils"
+	"github.com/crypto-power/cryptopower/ui/values"
+)
+
+const (
+	// These are constants used to represent various rate sources supported.
+	bittrex = values.BittrexExchange
+	binance = values.BinanceExchange
+	none    = values.DefaultExchangeValue
+
+	// These are method names for expected bittrex specific websocket messages.
+	BittrexMsgHeartbeat  = "heartbeat"
+	BittrexMarketSummary = "marketSummary"
+	BittrexTicker        = "ticker"
+
+	// MktSep is used repo wide to separate market symbols.
+	MktSep = "-"
+)
+
+var (
+	// These are urls to fetch rate information from the Bittrex exchange.
+	bittrexURLs = sourceURLs{
+		price: "https://api.bittrex.com/v3/markets/%s/ticker",
+		stats: "https://api.bittrex.com/v3/markets/%s/summary",
+		// Bittrex uses SignalR, which retrieves the actual websocket endpoint
+		// via HTTP.
+		ws: "socket-v3.bittrex.com",
+	}
+
+	// These are urls to fetch rate information from the Binance exchange.
+	binanceURLs = sourceURLs{
+		// See: https://binance-docs.github.io/apidocs/spot/en/#current-average-price
+		price: "https://api.binance.com/api/v3/ticker/24hr?symbol=%s",
+		ws:    "wss://stream.binance.com:9443/stream?streams=%s",
+	}
+
+	// supportedMarkets is a map of markets supported by rate sources
+	// implemented (Binance, Bittrex).
+	supportedMarkets = map[string]*struct{}{
+		values.BTCUSDTMarket: {},
+		values.DCRUSDTMarket: {},
+		values.LTCUSDTMarket: {},
+		values.DCRBTCMarket:  {},
+		values.LTCBTCMarket:  {},
+	}
+
+	// binanceMarkets is a map of Binance formatted market to the repo's format,
+	// e.g BTCUSDT : BTC-USDT. This is to facilitate quick lookup and to/fro
+	// market name formatting.
+	binanceMarkets = make(map[string]string)
+
+	// Rates exceeding rateExpiry are expired and should be removed unless there
+	// was an error fetching a new rate.
+	rateExpiry = 30 * time.Minute
+
+	// Rate sources should be refreshed every RateRefreshDuration to replace
+	// expired rates and reconnect websocket if need be.
+	RateRefreshDuration = 60 * time.Minute
+
+	rateNotificationInterval = 5 * time.Minute
```

Comment:
@ukane-philemon can you confirm this works?, i don't think it does, updated now still happens every 2 seconds